### PR TITLE
feat(realy): Add Relay best practices guide

### PR DESCRIPTION
**Pull request by Dav1dde**
Closes: https://github.com/getsentry/team-ingest/issues/90

Some best practices for development on Relay.

## IS YOUR CHANGE URGENT?  

Help us prioritize incoming PRs by letting us know when the change needs to go live.
- [ ] Urgent deadline (GA date, etc.): 
- [ ] Other deadline: 
- [X] None: Not urgent, can wait up to 1 week+

## SLA

- Teamwork makes the dream work, so please add a reviewer to your PRs.
- Please give the docs team up to 1 week to review your PR unless you've added an urgent due date to it.
Thanks in advance for your help!

## PRE-MERGE CHECKLIST

*Make sure you've checked the following before merging your changes:*

- [x] Checked Vercel preview for correctness, including links
- [x] PR was reviewed and approved by any necessary SMEs (subject matter experts)
- [ ] PR was reviewed and approved by a member of the [Sentry docs team](https://github.com/orgs/getsentry/teams/docs)

### CODE REVIEW

File: develop-docs/ingestion/relay-best-practices.mdx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+Relay is a critical component in Sentry's infrastructure handling hundreds of thousands of requests per second.
+To make sure your changes to Relay go smoothly, there are some best practices to follow.
+
+
+## Forward Compatibility
+
+Make sure changes to the event protocol and APIs are forward-compatible. Relay should not drop
+or truncate data it does not understand. It is a supported use-case to have customers running
+outdated Relays but up-to-date SDKs.
+
```
Comment (jjbayer) [inline develop-docs/ingestion/relay-best-practices.mdx:14]: We could mention `retain=true` here.

File: develop-docs/ingestion/relay-best-practices.mdx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+## Data Structures
+
+All data structures come with a cost, consider using the correct data structure for your workload.
+
+- A `BTreeMap` may be more performant than a `HashMap` for small amounts of data.
+- Use `ahash` (through `hashbrown::HashMap`) as hasher for a `HashMap` to reduce the hashing overhead for high throughput
+`HashMap`'s.
+- Use `smallvec` over `Vec` when your data is likely to be small.
+- Consider using a priority queue or heap if you need sorted access to a large amount of data.
```
Comment (jjbayer) [inline develop-docs/ingestion/relay-best-practices.mdx:103]: nit

```suggestion
- If possible, avoid iterating through large collections in regular intervals (e.g. project configs or metric buckets). Instead, use a [priority queue](https://en.wikipedia.org/wiki/Priority_queue) or similar to access only the elements that are currently relevant.
```

File: develop-docs/ingestion/relay-best-practices.mdx [hunk 3]
```diff
@@ -0,0 +1,120 @@
+---
+title: Relay Best Practices
+---
+
+Relay is a critical component in Sentry's infrastructure handling hundreds of thousands of requests per second.
+To make sure your changes to Relay go smoothly, there are some best practices to follow.
```
Comment (jjbayer) [inline develop-docs/ingestion/relay-best-practices.mdx:6]: Could link to https://develop.sentry.dev/engineering-practices/rust/ here.

**Review by jjbayer — APPROVED** [review]
Thanks!

File: develop-docs/ingestion/relay-best-practices.mdx (hunk 1 above)
Comment (Dav1dde) [reply develop-docs/ingestion/relay-best-practices.mdx:17]: Not sure how to phrase this and if we start naming things we probably want to name all the tools available (e.g. `additional_properties`). Gonna leave this for a follow-up or maybe even an extra page.

**Pull request merged**